FLT/AutomorphicRepresentation/Example.lean

```diff
@@ -408,7 +408,79 @@ noncomputable abbrev unitszHatsub : Subgroup QHatˣ :=
 noncomputable abbrev unitszsub : Subgroup QHatˣ :=
   (Units.map (Int.castRingHom QHat : ℤ →* QHat)).range
 
-lemma unitsrat_meet_unitszHat : unitsratsub ⊓ unitszHatsub = unitszsub := sorry
+lemma unitsrat_meet_unitszHat : unitsratsub ⊓ unitszHatsub = unitszsub := by
+  apply le_antisymm
+  · intro x ⟨⟨q, hxq⟩, ⟨zHat, hxzHat⟩⟩
+    have h : Units.val x ∈ zsub := by
+      rw [← rat_meet_zHat]
+      constructor
+      · use Units.val q
+        simp [← hxq]
+      · use Units.val zHat
+        simp [← hxzHat]
+    obtain ⟨z, hz⟩ := h
+    have znez : z ≠ 0 := by
+      intro h
+      simp [h, Eq.comm] at hz
+    let a := Int.sign z
+    let b := Int.natAbs z
+    let zinvRat : ℚ := mkRat a b
+    have hzinvRat : ↑z * zinvRat = 1 := by
+      unfold zinvRat a b
+      rw [Rat.mkRat_eq_div, mul_div, ← Rat.intCast_mul, Int.mul_sign, Int.cast_natCast,
+      ← Rat.natCast_div_self, Nat.div_self (by simp [znez]), Nat.cast_one]
+    let zinvZHat : ZHatˣ := zHat⁻¹
+    have hzinvZHat : ↑zHat * ↑zinvZHat = (1 : ZHat) := by
+      rw [mul_comm, Units.mul_eq_one_iff_eq_inv]
+    let xinv : QHatˣ := x⁻¹
+    have h1 : zinvRat ⊗ₜ[ℤ] (1 : ZHat) = xinv := by
+      apply Units.eq_inv_of_mul_eq_one_left
+      rw [← hz, AddMonoidHom.coe_coe, eq_intCast, ← zsmul_eq_mul,
+      TensorProduct.smul_tmul', zsmul_eq_mul, hzinvRat, Algebra.TensorProduct.one_def]
+    have h2 : (1 : ℚ) ⊗ₜ[ℤ] (Units.val zinvZHat) = xinv := by
+      apply Units.eq_inv_of_mul_eq_one_left
+      have hzHat : (1 : ℚ) ⊗ₜ[ℤ] (zHat : ZHat) = (x : QHat) := by simp [← hxzHat]
+      rw [← hzHat, Algebra.TensorProduct.tmul_mul_tmul, mul_one, hzinvZHat, Algebra.TensorProduct.one_def]
+    have h3 : zinvRat ⊗ₜ[ℤ] (1 : ZHat) = (1 / b : ℚ) ⊗ₜ[ℤ] (a : ZHat) := by
+      unfold zinvRat
+      rw [Rat.mkRat_eq_div, ← mul_one (@Int.cast ℚ Rat.instIntCast a), ← mul_div,
+      ← zsmul_eq_mul, TensorProduct.smul_tmul, zsmul_eq_mul, mul_one]
+    have bpos : 0 < b := Int.natAbs_pos.2 znez
+    have heq : (1 / (((Nat.toPNat b bpos) : ℕ) : ℚ)) ⊗ₜ[ℤ] (a : ZHat) = (1 / (((1 : ℕ+) : ℕ) : ℚ)) ⊗ₜ[ℤ] ↑zinvZHat := by
+      have : ↑(Nat.toPNat b bpos) = b := by
+        unfold Nat.toPNat
+        rw [PNat.mk_coe]
+      rw [PNat.val_ofNat, Nat.cast_one, div_self one_ne_zero, this, ← h3, h1, h2]
+    have cop1 : IsCoprime (b.toPNat bpos) ↑a := by
+      rw [IsCoprime, ZHat.intCast_val, ← ZMod.isUnit_natAbs,
+        ZMod.isUnit_iff_coprime, Int.natAbs_sign_of_nonzero znez]
+      exact Nat.coprime_one_left _
+    have cop2 : IsCoprime 1 ↑zinvZHat := by
+      simp only [IsCoprime, PNat.val_ofNat]
+      exact isUnit_of_subsingleton _
+    obtain ⟨hb, ha⟩ := (lowestTerms ↑x).2 (Nat.toPNat b bpos) 1 ↑a ↑zinvZHat ⟨cop1, cop2, heq⟩
+    have b1 : b = 1 := by
+      have : (1 : ℕ) = ↑(@OfNat.ofNat ℕ+ 1 (instOfNatPNatOfNeZeroNat 1)) := by norm_cast
+      rw [this, ← hb]
+      norm_cast
+    apply Int.isUnit_iff_natAbs_eq.2 at b1
+    obtain ⟨u, hu⟩ := b1
```

Comment:
```suggestion
    obtain ⟨u, rfl⟩ := Int.isUnit_iff_natAbs_eq.2 b1
```
no need to change `b1` and life is easier if z is just defined to be `↑u`; then all the `z`s are completely eliminated.